optuna/_optimize.py

```diff
@@ -193,6 +195,19 @@ def _run_trial(
     func_err_fail_exc_info: Optional[Any] = None
     # Set to a string if `func` returns correctly but the return value violates assumptions.
     values_conversion_failure_message: Optional[str] = None
+    stop_event: Optional[threading.Event] = None
+    thread: Optional[threading.Thread] = None
+
+    if (
+        isinstance(study._storage, storages._CachedStorage)
```

Comment:
Should we account for `RDBStorage` instances and perhaps general storages with heartbeat support?

Reply:
I added `BaseStorage.{check_heartbeat_support, get_heartbeat_interval}`. This reduces conditioning on specific storage classes.